resources/aws_ec2.go

```diff
@@ -663,3 +714,365 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
 func (obj *AwsEc2Res) prependName() string {
 	return AwsPrefix + obj.GetName()
 }
+
+// snsServer returns an http server used to listen for sns messages
+func (obj *AwsEc2Res) snsServer() *http.Server {
+	port := fmt.Sprintf(":%d", int(obj.SnsPort))
+	srv := &http.Server{Addr: port}
+	log.Printf("%s: Starting SNS Endpoint", obj)
+	//rand.Seed(time.Now().UTC().UnixNano())
+	http.HandleFunc("/", obj.postHandler)
+	log.Printf("%s: Listening on %s", obj, port)
+	go func() {
+		if err := srv.ListenAndServe(); err != nil {
+			select {
+			case obj.awsChan <- &chanStruct{
+				err: errwrap.Wrapf(err, "error starting http server"),
+			}:
+			case <-obj.closeChan:
+			}
+			return
+		}
+	}()
+	return srv
+}
+
+// postHandler listens for posts on the SNS Endpoint
+// TODO: download pem and check message against signature
+func (obj *AwsEc2Res) postHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method == "POST" {
+		post, _ := ioutil.ReadAll(req.Body)
+		// unmarshall json
+		type data struct {
+			Type    string
+			Token   string
+			Message string
+		}
+		var p data
+		err := json.Unmarshal(post, &p)
+		if err != nil {
+			select {
+			case obj.awsChan <- &chanStruct{
+				err: errwrap.Wrapf(err, "error unmarshaling policy"),
+			}:
+			case <-obj.closeChan:
+			}
+			return
+		}
+		if p.Type == "SubscriptionConfirmation" {
+			csInput := &sns.ConfirmSubscriptionInput{
+				Token:    aws.String(p.Token),
+				TopicArn: aws.String(obj.topicArn),
+			}
+			_, err := obj.snsClient.ConfirmSubscription(csInput)
+			if err != nil {
+				select {
+				case obj.awsChan <- &chanStruct{
+					err: errwrap.Wrapf(err, "error confirming subscription"),
+				}:
+				case <-obj.closeChan:
+				}
+				return
+			}
+			log.Printf("%s: Subscription Confirmed", obj)
+			err = obj.Running()
+			if err != nil {
+				select {
+				case obj.awsChan <- &chanStruct{
+					err: errwrap.Wrapf(err, "error calling obj.Running()"),
+				}:
+				case <-obj.closeChan:
+				}
+				return
+			}
+		}
+		if p.Type == "Notification" {
+			event := p.Message
+			select {
+			case obj.awsChan <- &chanStruct{
+				str: event,
+			}:
+			case <-obj.closeChan:
+				return
+			}
+		}
+	} else {
+		http.Error(w, "Invalid request method", http.StatusMethodNotAllowed)
+	}
+}
+
+// makeSnsTopic creates a topic on aws sns
+func (obj *AwsEc2Res) makeSnsTopic() {
```

Comment:
this is basically a method with a bunch of side effects... Why isn't it a regular method that returns `error` ?
In the caller we can these do this `select` stuff or not, depending on the way it's called and if needed...
same comment for a few other functions...

Reply:
To clarify, return the error and do if err = makeSnsTopic() { handle error } where it's called?

Reply:
I think that pattern would be more correct. It also means each of these small worker functions have fewer side effects, and the large error handling sections using select would be de-duplicated.